core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -728,7 +800,22 @@ class BlockManagerMasterEndpoint(
     }
 
     if (storageLevel.isValid) {
+      val firstBlock = locations.isEmpty
       locations.add(blockManagerId)
+
+      blockId.asRDDId.foreach { rddBlockId =>
+        (trackingCacheVisibility, firstBlock) match {
+          case (true, true) =>
+            // Mark as invisible for the first block.
+            invisibleRDDBlocks.add(rddBlockId)
+          case (true, false) if !invisibleRDDBlocks.contains(rddBlockId) =>
+            // If the rdd block is already visible, ask storage manager to update the visibility
+            // status.
+            blockManagerInfo(blockManagerId).storageEndpoint
+              .ask[Unit](MarkRDDBlockAsVisible(rddBlockId))
```

Comment:
To understand better, this is to handle case where block was replicated to another node - but in meantime the original task completed successfully (and so not invisible anymore).
So ensure that the replica is also marked as visible, right ?

Any other scenario where this is relevant ?

Reply:
> So ensure that the replica is also marked as visible, right ?

Yes, this is one scenario. Another scenario is that once a task failed fetching the cached block from remote executor, it'll compute and cache the block again.

Reply:
Ah yes, good point. Makes sense to do this irrespective of the path to block creation.